### PR TITLE
Enhanced cloud provider settings and credentials validation/testing.

### LINKED ISSUE #209: Validate AWS Credentials.
Add validation in the storage controller to validate/test AWS Credentials by creating a _test_ session.
Set a condition on error.

### PR DESCRIPTION
**Pull request by jortel**
fixes #209

Enhance validation of `MigStorage` backup storage and volume snapshot settings.  This includes inspecting the _credentials_ secret and running _test_ operations to validate the credentials and settings.  To validate backup storage settings and credentials, The _AWSProvider_ uploads a 1 byte file to the _bucket_ followed by a download and delete.  For volume snapshot storage, it creates and deletes a small volume.  All of this runs very quickly.  

As part of this effort, it seemed valuable to promote a _cloud provider_ to a 1st class concept.  Doing so centralizes provider related functionality.  As a result, multiple places in the code that had _switch_ statements on the provider type could be simplified by delegating to the new cloud _Provider_ object.  The provider supports validation, creating/updating velero related resources, and testing against actual provider using settings and credentials. This resulted in a bit more code churn than I would have liked but believe it is worth it in the end.

The PR includes (2) new dependencies.

### CODE REVIEW

File: pkg/cloudprovider/azure.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	ResourceGroup  string
+	APITimeout     string
+}
+
+func (p *AzureProvider) UpdateBSL(bsl *velero.BackupStorageLocation) {
+	bsl.Spec.Provider = Azure
+}
+
+func (p *AzureProvider) UpdateVSL(vsl *velero.VolumeSnapshotLocation) {
+	vsl.Spec.Provider = AWS
```
Comment (sseago) [inline pkg/cloudprovider/azure.go]: Shouldn't this be Azure as well?

File: pkg/cloudprovider/gcp.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	velero "github.com/heptio/velero/pkg/apis/velero/v1"
+	kapi "k8s.io/api/core/v1"
+)
+
+type GCPProvider struct {
+	BaseProvider
+}
+
+func (p *GCPProvider) UpdateBSL(bsl *velero.BackupStorageLocation) {
+	bsl.Spec.Provider = Azure
```
Comment (sseago) [inline pkg/cloudprovider/gcp.go]: GCP?

File: pkg/cloudprovider/azure.go (hunk 1 above)
Comment (jortel) [reply pkg/cloudprovider/azure.go]: _cut-n-paste_ error.  Fixed.

File: pkg/cloudprovider/gcp.go (hunk 2 above)
Comment (jortel) [reply pkg/cloudprovider/gcp.go]: _cut-n-paste_ error.  Fixed.

**Pull request merged**